src/nnbench/context.py

```diff
@@ -183,3 +187,256 @@ def __call__(self) -> dict[str, Any]:
+        """
+        Flatten the context's dictionary, converting nested dictionaries into a single dictionary with keys separated by `sep`.
+
+        Parameters
+        ----------
+        prefix : str, optional
+            A prefix to prepend to each key in the flattened dictionary.
+        sep : str, optional
+            The separator used to join nested keys.
+        inplace : bool, optional
+            If True, the current context's dictionary is modified in-place. Otherwise, return a new Context instance.
+
+        Returns
+        -------
+        Context
+            The current context instance if `inplace` is True; otherwise, a new, flattened Context instance.
+        """
+
+        if inplace:
+            self._ctx_dict = self.flatten_dict(d=self._ctx_dict, prefix=prefix, sep=sep)
+            return self
+        else:
+            return Context(context=self.flatten_dict(self._ctx_dict, prefix=prefix, sep=sep))
+
+    @staticmethod
+    def unflatten_dict(d: dict[str, Any], sep: str = ".") -> dict[str, Any]:
+        """
+        Recursively unflatten a dictionary by expanding keys seperated by `sep` into nested dictionaries.
+
+        Parameters
+        ----------
+        d : dict[str, Any]
+            The dictionary to unflatten.
+        sep : str, optional
+            The separator used in the flattened keys.
+
+        Returns
+        -------
+        dict[str, Any]
+            The unflattened dictionary.
+        """
+        sorted_keys = sorted(d.keys())
+        unflattened = {}
+        for prefix, keys in itertools.groupby(sorted_keys, key=lambda key: key.split(sep, 1)[0]):
+            key_group = list(keys)
+            if len(key_group) == 1 and sep not in key_group[0]:
+                unflattened[prefix] = d[prefix]
+            else:
+                nested_dict = {key.split(sep, 1)[1]: d[key] for key in key_group}
+                unflattened[prefix] = Context.unflatten_dict(d=nested_dict, sep=sep)
+        return unflattened
+
+    def unflatten(self, sep: str = ".", inplace: bool = True) -> "Context":
+        """
+        Unflatten the context's dictionary, expanding keys into nested dictionaries on the `sep` value.
+
+        Parameters
+        ----------
+        sep : str, optional
+            The separator used in the flattened keys.
+        inplace : bool, optional
+            If True, the current context's dictionary is modified in-place. Otherwise, return a new Context instance.
+
+        Returns
+        -------
+        Context
+            The current context instance if `inplace` is True; otherwise, a new unflattened Context instance.
+        """
+        if inplace:
+            self._ctx_dict = self.unflatten_dict(self._ctx_dict, sep=sep)
+            return self
+        else:
+            return Context(context=self.unflatten_dict(self._ctx_dict, sep=sep))
+
+    @staticmethod
+    def filter_dict(d: dict[str, Any], predicate: Callable[[str, Any], bool]) -> dict[str, Any]:
```

Comment:
And again, though I think this might be a deletion candidate once we implement nested key/value views.